pkg/email/email_test.go

```diff
@@ -0,0 +1,89 @@
+package email_test
+
+import (
+	"errors"
+	"html/template"
+	"testing"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/email"
+	"github.com/treeverse/lakefs/pkg/testutil"
+)
+
+func TestNewEmailer_DialerAndLimiter(t *testing.T) {
+	p := email.Params{}
+	emailer, err := email.NewEmailer(p)
+	testutil.Must(t, err)
+	if emailer.Dialer == nil {
+		t.Errorf("got nil email dialer")
+	}
+	if emailer.Limiter == nil {
+		t.Errorf("got nil email limiter")
+	}
+}
+
+func TestNewEmailer_Sender(t *testing.T) {
+	p := email.Params{
+		SMTPHost: "foo",
+		Sender:   "bar",
+	}
+	_, err := email.NewEmailer(p)
+	if err == nil {
+		t.Errorf("got a proper emailer, expected %s", email.ErrNoSenderConfigured)
+	}
+}
+
+func TestSendEmail(t *testing.T) {
+	p := email.Params{}
+	emailer, err := email.NewEmailer(p)
+	testutil.Must(t, err)
+	err = emailer.SendEmail([]string{"foo"}, "bar", "baz", nil)
+	if !errors.Is(err, email.ErrNoSMTPHostConfigured) {
+		t.Errorf("expected err %s", email.ErrNoSMTPHostConfigured)
+	}
+	emailer.Params.SMTPHost = "foo"
+	err = emailer.SendEmail([]string{"foo"}, "bar", "baz", nil)
+	if !errors.Is(err, email.ErrNoSenderConfigured) {
+		t.Errorf("email sent, expected err %s", email.ErrNoSenderConfigured)
+	}
+	emailer.Params.Sender = "bar"
+	err = emailer.SendEmail([]string{}, "bar", "baz", nil)
+	if !errors.Is(err, email.ErrNoRecipientConfigured) {
+		t.Errorf("email sent, expected err %s", email.ErrNoRecipientConfigured)
+	}
+	err = emailer.SendEmail([]string{"foo"}, "bar", "baz", nil)
+	if err == nil {
+		t.Errorf("email sent, expected error: dial tcp :0: connect: can't assign requested address")
+	}
+}
+
+func TestSendEmailWithLimit(t *testing.T) {
+	p := email.Params{Burst: 0, LimitEveryDuration: time.Minute}
+	emailer, err := email.NewEmailer(p)
+	testutil.Must(t, err)
+	err = emailer.SendEmailWithLimit([]string{"foo"}, "bar", "baz", nil)
+	if !errors.Is(err, email.ErrRateLimitExceeded) {
+		t.Errorf("expected error %s", email.ErrRateLimitExceeded)
+	}
+}
+
+func TestBuildEmailByTemplate(t *testing.T) {
```

Comment:
1. You exposed a private function for testing - which is a valid option - but you cloud also check the public functionality, which is usually better in order not to couple of the test with the specific implementation.
2. In addition to the public function you exposed the internal templates and also tested with with a set of other testCases. I think it will be better testing the expectation from the template should be tested and not the actual template. And in case you are testing the actual template, put it in the same test cases or explain what you are testing in each test case by write table driven tests.

Reply:
1. Changed the function to private.
2. Added a function that tests the expected template with test tables